### ISSUE TITLE
Inference on empty bounds range

### ISSUE THREAD

**Issue by saeednj** [microsoft/checkedc-clang#697]
In the following code an array with bounds `R1: (a, a + 10)` is declared, but passed to a function `foo` with a negative element count which generates bounds `R2: (a, a - 5)`. This code compiles without any errors or warnings. The injected dynamic check is triggered at runtime though.
```c
#include <stdchecked.h>

int foo(array_ptr<int> a : count(n), int n)
{
    return a[n / 2];
}

int main()
{
    int i;
    int a checked[10];
    for( i=0; i<10; i++ ) a[i] = i;

    return foo(a, -5);
}
```
In the bounds inference logic, `R2` should be a sub-range of `R1` to be acceptable. Although the offsets are constant values, the compiler does not catch that `R2` is empty and therefore any access to `a` is a violation of bounds. 

We can handle this case by adding this snippet to the `BaseRange::InRange` function in `SemaBounds.cpp`:
```c
          if ( R.IsEmpty() ) {
            Cause = CombineFailures(Cause, ProofFailure::Empty);
            return ProofResult::False;
          }
```
The case of source bounds (`R1`) being empty seems to be handled by the code already, but should be tested.
